### ISSUE TITLE
KL divergence

### ISSUE THREAD

**Issue by luca310795** [kumar-shridhar/PyTorch-BayesianCNN#28]
Hi,
in BBBlayers.py you compute KL[p || q], but according to the ELBO formula you should compute  KL[q || p]. Why is it like that? Thanks!

**Comment by kumar-shridhar** [comment]
Hi @luca310795 ,

KL divergence (p(x), q(x)), is a measure of the information lost when q(x) is used to approximate p(x). So thats why we calculated the KL (p(x), q(x)) in BBBlayers.py.

**Comment by Zimba96** [comment]
Hi @kumar-shridhar,
I believe @luca310795 is right on that point. I was also wondering about your implementation of the KL divergence since other git-projects dealing with that topic like e. g. [JavierAntoran's](https://github.com/JavierAntoran/Bayesian-Neural-Networks) use it exactly the other way around. As far as I know, the KL divergence is non-symmetric, meaning KL[q || p] != KL[p || q]. According to [Blundell et al.](https://arxiv.org/pdf/1505.05424.pdf) we need KL[q || p] where q is our approximation to the prior distribution p for our loss function. According to [this thread](https://stats.stackexchange.com/questions/234757/how-to-use-kullback-leibler-divergence-if-mean-and-standard-deviation-of-of-two) which explains the computation of the KL divergence between two distributions (I believe they're calculating KL[p || q] there!) given their parameters, your calculation in BBBlayers.py would therefore be KL[p || q] instead of the required KL[q || p].
Hence, I believe that the calculation in the BBBlayers should actually be 
`kl = KL_DIV(self.W_mu, selfW_sigma, self.prior_mu, self.prior_sigma)`
as we would like to approximate our weights-distribution (which would be q) towards our prior distribution (which would be p).
If I'm wrong about something, please explain.
Thanks!

**Comment by Nebularaid2000** [comment]
@Zimba96 I have the same question regarding this part. I think your explanation is correct. It should be
`kl = KL_DIV(self.W_mu, self.W_sigma, self.prior_mu, self.prior_sigma)`
